docs/cryptodoc/src/05_10_cmce.rst

```diff
@@ -0,0 +1,469 @@
+
+Classic McEliece's key generation algorithm utilizes its Goppa code, defined by a
+field ordering and a monic irreducible polynomial, to create a binary parity
+check matrix :math:`H` in systematic form. This matrix is represented by the
+``Classic_McEliece_Matrix`` class.
+
+The class follows the process outlined in Section 7.2 of [CMCE-ISO]_ for matrix
+creation. Initially, a binary :math:`mt \times n` matrix is created as
+described in Steps 1 and 2 of Section 7.2.2. Each row of the matrix is
+represented as a ``bitvector`` object. Subsequently, a constant-time Gauss
+algorithm is applied to reduce the matrix to the systematic form :math:`H=(I_{mt}|T)`.
+The algorithm achieves this by systematically applying XOR operations on pairs
+of matrix rows, resulting in the identity matrix on the left. Finally, the
+submatrix `T` is stored in the matrix object, mirroring its representation in
+the public key specified in Section 9.2.7 of [CMCE-ISO]_.
+
+Classic McEliece instances with the suffix ``f`` employ a semi-systematic transformation
+strategy, utilizing the parameters :math:`(\mu, \nu) = (32, 64)`, as described
+in Section 7.2.3 of [CMCE-ISO]_. Following the recommendation in Section 6.1 of
+the implementation guide accompanying [CMCE-R4]_, a Gauss algorithm is executed
+to create an identity matrix for the first :math:`mt-\mu` rows. Subsequently, a
+modified Gauss algorithm achieves a reduced row-echelon form for
+the :math:`\mu \times \nu` submatrix at position :math:`(\mu, \mu)`.
+This process determines the column selection, i.e., the indices of the non-zero
+columns. As Section 7.2.3 Step 5 of [CMCE-ISO]_ outlines, the matrix's
+columns and the field ordering are permuted according to the pivots. The main
+Gauss algorithm concludes at this point, leaving the matrix in its modified
+and systematic form.
+
+For encapsulation, the matrix :math:`H` is multiplied by an error vector
+:math:`e`. This multiplication is performed by computing the parity of the
+Hamming weight of :math:`e\ \textrm{XOR}\ r` for each row :math:`r` of :math:`H`.
+Matrix creation and multiplication are implemented in constant time for matrices
+with fixed dimensions.
+
+
+.. _pubkey/cmce/keys_internal:
+
+Key Pair
+^^^^^^^^
+
+Botan's key pair for Classic McEliece consists of two classes:
+``Classic_McEliece_PrivateKeyInternal`` and ``Classic_McEliece_PublicKeyInternal``.
+As defined in Section 9.2.12, the private key stores the key generation seed,
+column selection, monic irreducible polynomial, field ordering control bits,
+and the seed for implicit rejection. On the other hand, the public key
+contains the sub-matrix :math:`T` of the binary parity check matrix
+:math:`H = (I_{mt}|T)`.
+
+The class ``Classic_McEliece_KeyPair_Internal`` holds both the private and public
+keys and contains key generation method. Details are discussed in Section
+:ref:`Key Generation <pubkey/cmce/key_gen>`.
+
+
+.. _pubkey/cmce/encaps_internal:
+
+Encapsulation Internals
```

Comment:
There's this loop going to MAX_ATTEMPTS in raw_kem_encrypt to prevent endless loops. Should we mention this and the corresponding failure probabilities here?

Reply:
Yeah, good point!